### PR TITLE
HOSTEDCP-1766: Add Hypershift e2e on AKS

### PR DESCRIPTION
**Pull request by Patryk-Stefanski**

### CODE REVIEW

**Comment by Patryk-Stefanski** [comment]
/test all 
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks
/test all

**Comment by fxierh** [comment]
It seems this conflicts with https://github.com/openshift/release/pull/53560/files. Would it be possible to merge that PR first (it still needs approval) ?. Thank you!

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/test all 
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/test all
//pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.16-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/test all
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/test all
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks
/test all

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (fxierh) [inline ci-operator/step-registry/aks/provision/aks-provision-commands.sh:53]: Would the command fail when these args are null strings ?

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (fxierh) [inline ci-operator/step-registry/aks/provision/aks-provision-commands.sh:12]: It might be helpful to define quota slices for the `hypershift` clusterprofile in the `{region}--hypershift-quota-slice-{idx}` similar to the following:
https://github.com/openshift/release/blob/a2003bca867eaafadf2fc33e6ddee45b41d99de0/core-services/prow/02_config/_boskos.yaml#L725-L757

So we could leverage `$LEASED_RESOURCE` without having to pass an extra environment variable. 

But for now I guess we could keep the current implementation.

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (fxierh) [inline ci-operator/step-registry/aks/provision/aks-provision-commands.sh:27]: I'm wondering what happens if `--min-count` and/or `--max-count` are specified when `ENABLE_CLUSTER_AUTOSCALER` is false. 

No objections to the current implementation though.

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-fips-guest-f7

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (Patryk-Stefanski) [reply ci-operator/step-registry/aks/provision/aks-provision-commands.sh:27]: They're all required so the command would fail if any of them were missing.

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (Patryk-Stefanski) [reply ci-operator/step-registry/aks/provision/aks-provision-commands.sh:12]: Hmmm we dont use these anywhere within the hypershift e2e afaik.  @enxebre is this something we would like to explore if so I can look into it.

File: ci-operator/step-registry/aks/provision/aks-provision-commands.sh
Comment (Patryk-Stefanski) [reply ci-operator/step-registry/aks/provision/aks-provision-commands.sh:53]: No should proceed successfully, I will run one of your rehearsal jobs just to double check.

**Comment by fxierh** [comment]
> [ci-operator/step-registry/azure/OWNERS](https://github.com/openshift/release/blob/master/ci-operator/step-registry/azure/OWNERS)

/cc @jianlinliu

File: ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh
Comment (bryan-cox) [inline ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh:26]: @fxierh - didn't you have a way of extracting these from the release image? We should do that rather than hardcoding them here.

Comment (Patryk-Stefanski) [reply ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh:26]: yeah there is a command for retrieving these I will update the pr now https://github.com/openshift/release/blob/6241daf293a2d892960aabb6dc928cc4b1a28d1d/ci-operator/step-registry/cucushift/hypershift-extended/k8s-mgmt/get-guest-annotations/cucushift-hypershift-extended-k8s-mgmt-get-guest-annotations-commands.sh

Comment (fxierh) [reply ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh]: Oh, I missed that somehow. Thanks for bringing it up @bryan-cox !

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

File: ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh
Comment (fxierh) [reply ci-operator/step-registry/hypershift/azure/run-e2e/hypershift-azure-run-e2e-commands.sh]: There's a catch though: the 4.17 images e.g. registry.redhat.io/redhat/certified-operator-index:v4.17 do not exist yet. We may use the 4.16 images for now and switch to the 4.17 images when they become available.

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by fxierh** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-azure-aks-hypershift-fips-guest-f7

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse pull-ci-openshift-hypershift-release-4.17-e2e-aks

**Comment by Patryk-Stefanski** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.16-amd64-nightly-azure-aks-hypershift-fips-guest-f7

**Comment by fxierh** [comment]
/cc @jianlinliu @liangxia 

Could you PTAL ? Thank you

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__amd64-nightly.yaml
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__amd64-nightly.yaml:1273]: The 4.17 job is using 4.16 image version, is that on purpose?

Comment (fxierh) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__amd64-nightly.yaml:1273]: Yup. The 4.17 catalog images will be available only after this minor version goes GA. Until then, OpenShift clusters are installed using catalog images of version x.(y-1).

File: ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml
Comment (jianlinliu) [inline ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: I am thinking if it is worth to write a separate step for hypershift case.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__amd64-nightly.yaml
Comment (Patryk-Stefanski) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.17__amd64-nightly.yaml:1273]: yes https://github.com/openshift/release/pull/53588#discussion_r1670750815

File: ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml
Comment (Patryk-Stefanski) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: I can create a separate step, although this avoids duplicating code. if you feel strongly that this should be a separate step, please lmk.

Comment (jianlinliu) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: Sounds not a big problem to avoid duplicating codes, cc @jinyunma for double confirm.

Comment (jinyunma) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: looks good to me.
But if more pre-configuration steps need to be updated for hypershift testing(mount hypershift azure credentials and change to use this creds in the script), I guess it is deserved to create new cluster profile for the test, then we can reuse those steps directly without any change.

Comment (fxierh) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: @Patryk-Stefanski 

Had a discussion with @jinyunma. 

Ideally, we could add a key named `osServicePrincipal.json` to the `cluster-secrets-hypershift` secret in the vault. The content will be the Azure credentials used in the job, specifically the ones stored in the `hypershift-ci-jobs-azurecreds` secret. This approach allows us to obtain the credentials from `${CLUSTER_PROFILE_DIR}/osServicePrincipal.json`, thereby avoiding the need to mount the `hypershift-ci-jobs-azurecreds` secret to all steps involved in hypershift Azure jobs.

However, if the above solution is not feasible for now for any reason, this PR can be merged as is. The secret mount issue can be addressed later.

Comment (jianlinliu) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: Yes, ideally `osServicePrincipal.json` can be gotten dynamically which is bound with the `cluster_profile`, no need to mount manually the credential.

Comment (Patryk-Stefanski) [reply ci-operator/step-registry/azure/provision/resourcegroup/azure-provision-resourcegroup-ref.yaml:12]: yeah, I don't see any issues with the suggested approach but with that being said we use the same approach of mounting cred secrets with our aws jobs. I'd like to run this by @sjenning , but he's on PTO till Monday. In the meantime could we merge this without the change and if Seth approves of the new approach then we can address it in a separate pr next week?

**Comment by Patryk-Stefanski** [comment]
@liangxia could you approve this pr please

**Review by liangxia — APPROVED** [review]

**Pull request merged**
